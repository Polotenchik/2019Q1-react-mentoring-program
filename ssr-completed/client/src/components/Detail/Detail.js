import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import styled, { css } from 'styled-components';
import Theme from '../../common/Theme/Theme';
import Logo from '../../common/Components/Logo/Logo';
import HeaderBackground from '../../common/Components/HeaderBackground/HeaderBackground';
import HeaderWrapper from '../../common/Components/HeaderWrapper/HeaderWrapper';
import { fetchMovie, clearState } from '../../store';

class Detail extends Component {

  componentWillMount() {
    const { match, fetchMovie } = this.props;
    const { id } = match.params;
    fetchMovie(id);
  }

  componentDidUpdate(prevProps) {
    const { fetchMovie, match } = this.props;
    if (match.params.id !== prevProps.match.params.id) {
      const { id } = match.params;
      fetchMovie(id);
    }
  }

  render() {
    const { targetFilm, clearState } = this.props;
    if (!targetFilm) {
      return (
        <EmptyWrapper>Loading...</EmptyWrapper>
      );
    }
    return (
      <HeaderBackground>
        <HeaderWrapper>
          <Container>
            <Logo />
            <ButtonLink  to="/" onClick={clearState} id="backToSearch">Search</ButtonLink>
          </Container>
          <Container>
            <PosterWrapper>
              <Poster src={targetFilm.poster_path} />
            </PosterWrapper>
            <InfoWrapper>
              <Title>{targetFilm.title}</Title>
              <Info>
                {targetFilm.genres.join(', ')}
              </Info>
              <div>
                <Info bold>
                  {targetFilm.release_date.split('-')[0]}
                </Info>
                <Info bold>
                  {targetFilm.runtime}
                  {'min'}
                </Info>
              </div>
              <Info>
                {targetFilm.overview}
              </Info>
            </InfoWrapper>
          </Container>
        </HeaderWrapper>
      </HeaderBackground>
    );
  }
}

export default withRouter(connect(({ movie }) => ({
  targetFilm: movie.activeFilm,
}), {
  fetchMovie, clearState,
})(Detail));

Detail.propTypes = {
  targetFilm: PropTypes.shape({}),
  fetchMovie: PropTypes.func,
  match: PropTypes.shape({}),
};

const Container = styled.div`
    display: flex;
    justify-content: space-between; 
    
    ${props => props.column && css`
        flex-direction: column;
    `}   
`;

const PosterWrapper = styled.div`
    position: relative;
    width: 15%;
    height: 0;
    padding-top: 22.5%;
`;

const InfoWrapper = styled.div.attrs({
  className: 'film-info-wrapper',
})`
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    width: 80%;        
`;

const Poster = styled.img`
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;   
`;

const Title = styled.h3`
    color: ${Theme.colors.red};  
    font-size: 24px;
    margin: 10px 0;  
`;

const Info = styled.span`
    color: ${Theme.colors.white};  
    font-weight: 400;
    padding: 10px;
    
     ${props => props.bold && css`
        font-weight: 800;
    `}
`;

const ButtonLink = styled(Link)`
    background-color: ${Theme.colors.white};
    color: ${Theme.colors.red};
    text-decoration: none;
    padding: 6px;
    border-radius: 6px;
`;

const EmptyWrapper = styled.div`
    display: flex;
    justify-content: center;
    color: ${Theme.colors.red};
    font-size: 24px;
    flex: auto;
`;
