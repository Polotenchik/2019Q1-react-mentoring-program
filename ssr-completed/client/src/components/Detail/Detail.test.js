import React from 'react';
import { shallow } from 'enzyme';
import Detail from './Detail';

describe('Detail Snapshot', () => {
    it('Deatil Component should render correctly with given props', () => {
        const targetFilm={
            "id":423,
            "title":"The Pianist",
            "tagline":"Music was his passion. Survival was his masterpiece.",
            "vote_average":8.1,
            "vote_count":2571,
            "release_date":"2002-09-24",
            "poster_path":"https://image.tmdb.org/t/p/w500/iunmxWkOi7Vk17Ob3G2HwwjgHsr.jpg",
            "overview":"The true story of pianist Wladyslaw Szpilman's experiences in Warsaw during the Nazi occupation. When the Jews of the city find themselves forced into a ghetto, Szpilman finds work playing in a caf√©; and when his family is deported in 1942, he stays behind, works for a while as a laborer, and eventually goes into hiding in the ruins of the war-torn city.",
            "budget":35000000,
            "revenue":120072577,
            "genres":["Drama","War"],
            "runtime":150
        };
        const searchButtonHandler=jest.fn();
        const component = shallow(<Detail targetFilm={targetFilm} searchButtonHandler={searchButtonHandler} />);
        expect(component).toMatchSnapshot();
    });
});

