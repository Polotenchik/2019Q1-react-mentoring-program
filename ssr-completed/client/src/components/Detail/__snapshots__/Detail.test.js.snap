// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Snapshot Deatil Component should render correctly with given props 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <styled.div>
        <Logo />
        <styled.button
          id="backToSearch"
          onClick={[Function]}
          white={true}
        >
          Search
        </styled.button>
      </styled.div>
      <styled.div>
        <styled.div>
          <styled.img
            src="https://image.tmdb.org/t/p/w500/iunmxWkOi7Vk17Ob3G2HwwjgHsr.jpg"
          />
        </styled.div>
        <styled.div>
          <styled.h3>
            The Pianist
          </styled.h3>
          <styled.span>
            Drama, War
          </styled.span>
          <div>
            <styled.span
              bold={true}
            >
              2002
            </styled.span>
            <styled.span
              bold={true}
            >
              150
              min
            </styled.span>
          </div>
          <styled.span>
            The true story of pianist Wladyslaw Szpilman's experiences in Warsaw during the Nazi occupation. When the Jews of the city find themselves forced into a ghetto, Szpilman finds work playing in a caf√©; and when his family is deported in 1942, he stays behind, works for a while as a laborer, and eventually goes into hiding in the ruins of the war-torn city.
          </styled.span>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Films by  Drama  genre
    </styled.div>
  </styled.div>
</Fragment>
`;
